tough-kms/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+use tough::schema::key::{Key, RsaKey, RsaScheme};
+use tough::sign::Sign;
+
+/// Represents a Signing Algorithms for AWS KMS.
+#[derive(Debug, Clone, PartialEq)]
+pub enum KmsSigningAlgorithms {
+    /// The key type
+    Rsa(String),
+}
+
+/// Implements the `KeySource` trait for keys that live in AWS KMS
+pub struct KmsKeySource {
+    /// Identifies AWS account named profile, if not provided default AWS profile is used.
+    pub profile: Option<String>,
+    /// Identifies an asymmetric CMK in AWS KMS.
+    pub key_id: String,
+    /// KmsClient Object to query AWS KMS
+    pub client: Option<KmsClient>,
+}
+
+impl fmt::Debug for KmsKeySource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("KmsKeySource")
+            .field("key_id", &self.key_id)
+            .field("profile", &self.profile)
+            .finish()
+    }
+}
+
+/// Implement the `KeySource` trait.
+impl KeySource for KmsKeySource {
+    fn as_sign(
+        &self,
+    ) -> std::result::Result<Box<dyn Sign>, Box<dyn std::error::Error + Send + Sync + 'static>>
+    {
+        let kms_client = match self.client.to_owned() {
+            Some(value) => value,
+            None => client::build_client_kms(self.profile.as_deref())?,
+        };
+        // Get the public key from AWS KMS
+        let fut = kms_client.get_public_key(rusoto_kms::GetPublicKeyRequest {
+            key_id: self.key_id.clone(),
+            ..rusoto_kms::GetPublicKeyRequest::default()
+        });
+        let response = tokio::runtime::Runtime::new()
+            .context(error::RuntimeCreation)?
+            .block_on(fut)
+            .context(error::KmsGetPublicKey {
+                profile: self.profile.clone(),
+                key_id: self.key_id.clone(),
+            })?;
+        let pb_key: Decoded<RsaPem> = response
+            .public_key
+            .context(error::PublicKeyNone)?
+            .to_vec()
+            .into();
+        Ok(Box::new(KmsRsaKey {
+            profile: self.profile.clone(),
+            client: Some(kms_client.clone()),
+            key_id: self.key_id.clone(),
+            public_key: pb_key,
+            signing_algorithm: KmsSigningAlgorithms::Rsa(String::from("RSASSA_PSS_SHA_256")),
+        }))
+    }
+
+    fn write(
+        &self,
+        _value: &str,
+        _key_id_hex: &str,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        Ok(())
+    }
+}
+
+/// Implements the Sign trait for KMS rsa Key
+pub struct KmsRsaKey {
+    /// Key Id of Customer Managed Key in KMS used to sign the message
+    pub key_id: String,
+    /// Aws account profile
+    pub profile: Option<String>,
+    /// KmsClient Object to query AWS KMS
+    pub client: Option<KmsClient>,
+    /// Public Key corresponding to Customer Managed Key
+    pub public_key: Decoded<RsaPem>,
+    /// Signing Algorithm to be used for the Customer Managed Key
+    pub signing_algorithm: KmsSigningAlgorithms,
+}
```

Comment:
I don't think these fields should be `pub` — anything going on here is purely an implementation detail.

Reply:
Done.